src/data/matrix/kronecker_product_bialgebra.lean

```diff
@@ -28,33 +28,89 @@ namespace matrix_bialgebra
+  (hf : ∀ a b₁ b₂, f a (b₁ + b₂) = f a b₁ + f a b₂)
+  (A : matrix l m α) (B₁ B₂ : matrix n p β) :
+  kronecker_map f A (B₁ + B₂) = kronecker_map f A B₁ + kronecker_map f A B₂ :=
+ext $ λ i j, hf _ _ _
+
+lemma kronecker_map_smul_left [has_scalar R α] [has_scalar R γ] (f : α → β → γ)
+  (hf : ∀ (r : R) a b, f (r • a) b = r • f a b) (r : R)
+  (A : matrix l m α) (B : matrix n p β) :
+  kronecker_map f (r • A) B = r • kronecker_map f A B :=
+ext $ λ i j, hf _ _ _
+
+lemma kronecker_map_smul_right [has_scalar R β] [has_scalar R γ] (f : α → β → γ)
+  (hf : ∀ (r : R) a b, f a (r • b) = r • f a b) (r : R)
+  (A : matrix l m α) (B : matrix n p β) :
+  kronecker_map f A (r • B) = r • kronecker_map f A B :=
+ext $ λ i j, hf _ _ _
+
+@[simps]
+def kronecker_map_linear [comm_semiring R]
+  [add_comm_monoid α] [add_comm_monoid β] [add_comm_monoid γ]
+  [module R α] [module R β] [module R γ]
+  (f : α →ₗ[R] β →ₗ[R] γ) :
+  matrix l m α →ₗ[R] matrix n p β →ₗ[R] matrix (l × n) (m × p) γ :=
+linear_map.mk₂ R
+  (kronecker_map (λ r s, f r s))
+  (kronecker_map_add_left _ $ f.map_add₂)
+  (kronecker_map_smul_left _ $ f.map_smul₂)
+  (kronecker_map_add_right _ $ λ a, (f a).map_add)
+  (kronecker_map_smul_right _ $ λ r a, (f a).map_smul r)
+
+end
+
+
+variables {α R S β : Type*} [comm_semiring α] [comm_semiring R] [comm_semiring S] [comm_semiring β]
 variables [algebra α R] [algebra α S] [algebra α β] [algebra R β] [algebra S β]
 variables {l m n p l' m' n' p' : Type*}
 variables [fintype l] [fintype m] [fintype n] [fintype p]
 variables [fintype l'] [fintype m'] [fintype n'] [fintype p']
 
+variables (α β)
+
+-- TODO: move this
+def algebra.biprod [is_scalar_tower α R β] [is_scalar_tower α S β] :
+  R →ₗ[α] S →ₗ[α] β :=
+((algebra.lmul α β).to_linear_map.compl₂ ((algebra.linear_map S β).restrict_scalars α)).comp
+  ((algebra.linear_map R β).restrict_scalars α)
+
+@[simp]
+lemma algebra.biprod_apply [is_scalar_tower α R β] [is_scalar_tower α S β] (r : R) (s : S) :
+  algebra.biprod α β r s = algebra_map R β r * algebra_map S β s := rfl
+
+variables (R S β)
+
+@[simps]
 def kronecker_biprod [is_scalar_tower α R β] [is_scalar_tower α S β] :
   (matrix l m R) →ₗ[α] (matrix n p S) →ₗ[α] matrix (l × n) (m × p) β :=
-{ to_fun :=
-  begin
-    intro A,
-    use λ B, λ i j, (algebra_map R β (A i.1 j.1)) * (algebra_map S β (B i.2 j.2)),
-    all_goals {intros x y, ext},
-    { simp only [pi.add_apply, mul_add, ring_hom.map_add, dmatrix.add_apply] },
-    { simp only [pi.smul_apply],
-      rw [← is_scalar_tower.algebra_map_smul S x, id.smul_eq_mul, ring_hom.map_mul,
-        smul_def, (is_scalar_tower.algebra_map_apply α S β x).symm],
-      ring,
-      all_goals {exact is_scalar_tower.right} },
-  end,
-  map_add' := λ _ _, by {simp only [add_mul, ring_hom.map_add, dmatrix.add_apply], refl},
-  map_smul' := λ _ _, by {simp_rw [pi.smul_apply, ← smul_def, is_scalar_tower.smul_assoc],
-    refl},
-  }
+kronecker_map_linear $ algebra.biprod α β
```

Comment:
This then generalizes trivially to the `tensor_product` version by using `tensor_product.mk` here instead.